server/src/test/java/io/druid/server/listener/announcer/ListenerDiscovererTest.java

```diff
@@ -0,0 +1,231 @@
+
+    final String not_listen_key = "NOTlistener_key";
+    EasyMock.expect(serviceCacheBuilder.name(EasyMock.eq(not_listen_key))).andReturn(serviceCacheBuilderEmpty).once();
+    EasyMock.expect(serviceCacheBuilderEmpty.threadFactory(EasyMock.<ThreadFactory>anyObject()))
+            .andReturn(serviceCacheBuilderEmpty)
+            .once();
+    EasyMock.expect(serviceCacheBuilderEmpty.build()).andReturn(serviceCacheEmpty).once();
+    serviceCacheEmpty.start();
+    EasyMock.expectLastCall().once();
+    EasyMock.expect(serviceCacheEmpty.getInstances()).andReturn(ImmutableList.<ServiceInstance<Void>>of()).once();
+
+    EasyMock.reset(serviceDiscovery);
+    EasyMock.expect(serviceDiscovery.serviceCacheBuilder()).andReturn(serviceCacheBuilder).times(2);
+
+    EasyMock.replay(serviceDiscovery, serviceCacheBuilder, serviceCache, serviceCacheBuilderEmpty, serviceCacheEmpty);
+
+    final Collection<DruidNode> nodes = listenerDiscoverer.getNodes(listen_key);
+    final Collection<DruidNode> no_nodes = listenerDiscoverer.getNodes(not_listen_key);
+
+    EasyMock.verify(serviceDiscovery, serviceCacheBuilder, serviceCache, serviceCacheBuilderEmpty, serviceCacheEmpty);
+
+    Assert.assertEquals(1, nodes.size());
+    final DruidNode node = Iterables.getFirst(nodes, null);
+    Assert.assertNotNull(node);
+    Assert.assertEquals(host, node.getHost());
+    Assert.assertEquals(port, node.getPort());
+
+    Assert.assertTrue(no_nodes.isEmpty());
+  }
+
+
+  @Test
+  public void testStartError() throws Exception
+  {
+    final RuntimeException ex = new RuntimeException("test exception");
+    serviceDiscovery.start();
+    EasyMock.expectLastCall().andThrow(ex).once();
+    expectedException.expect(new BaseMatcher<Throwable>()
+    {
+      @Override
+      public boolean matches(Object o)
+      {
+        return o == ex;
+      }
+
+      @Override
+      public void describeTo(Description description)
+      {
+
+      }
+    });
+    EasyMock.replay(serviceDiscovery);
+    final ListenerDiscoverer listenerDiscoverer = new ListenerDiscoverer(serviceDiscovery);
+    listenerDiscoverer.start();
+    EasyMock.verify(serviceDiscovery);
+  }
+
+
+  @Test
+  public void testStopError() throws Exception
+  {
+    final IOException ex = new IOException("test exception");
+    serviceDiscovery.start();
+    EasyMock.expectLastCall().once();
+    serviceDiscovery.close();
+    EasyMock.expectLastCall().andThrow(ex).once();
+    expectedException.expectCause(new BaseMatcher<Throwable>()
+    {
+      @Override
+      public boolean matches(Object o)
+      {
+        return o == ex;
+      }
+
+      @Override
+      public void describeTo(Description description)
+      {
+
+      }
+    });
+    EasyMock.replay(serviceDiscovery);
+    final ListenerDiscoverer listenerDiscoverer = new ListenerDiscoverer(serviceDiscovery);
+    listenerDiscoverer.start();
+    listenerDiscoverer.stop();
+    EasyMock.verify(serviceDiscovery);
+  }
+
+  @Test(timeout = 5_000)
```

Comment:
problem is that curator has non-atomic / non-transactional stuff all in it that is easy to hit if you're not careful.

I *could* do a path watch on the expected zk path, and latch on that being removed, but it is not clear to me that approach is really better than what's here.